libs/ml/src/dataloaders/word2vec_loaders/w2v_dataloader.cpp

```diff
@@ -0,0 +1,398 @@
+  {
+    data_.push_back(std::move(indexes));
+    return true;
+  }
+  return false;
+}
+
+/**
+ *
+ * @tparam T
+ * @param filename
+ */
+template <typename T>
+void W2VLoader<T>::SaveVocab(std::string const &filename)
+{
+  vocab_.Save(filename);
+}
+
+/**
+ *
+ * @tparam T
+ * @param filename
+ */
+template <typename T>
+void W2VLoader<T>::LoadVocab(std::string const &filename)
+{
+  vocab_.Load(filename);
+}
+
+/**
+ * get size of the vocab
+ * @tparam T
+ * @return
+ */
+template <typename T>
+math::SizeType W2VLoader<T>::vocab_size() const
+{
+  return vocab_.GetVocabCount();
+}
+
+/**
+ * export the vocabulary by reference
+ * @tparam T
+ * @return
+ */
+template <typename T>
+typename W2VLoader<T>::VocabType const &W2VLoader<T>::vocab() const
+{
+  return vocab_;
+}
+
+/**
+ * helper method for retrieving a word given its index in vocabulary
+ * @tparam T
+ * @param index
+ * @return
+ */
+template <typename T>
+std::string W2VLoader<T>::WordFromIndex(SizeType index) const
+{
+  return vocab_.WordFromIndex(index);
+}
+
+/**
+ * helper method for retrieving word index given a word
+ * @tparam T
+ * @param word
+ * @return
+ */
+template <typename T>
+typename W2VLoader<T>::SizeType W2VLoader<T>::IndexFromWord(std::string const &word) const
+{
+  return vocab_.IndexFromWord(word);
+}
+
+template <typename T>
+typename W2VLoader<T>::SizeType W2VLoader<T>::window_size()
+{
+  return window_size_;
+}
+
+/**
+ * converts string to indices and inserts into vocab as necessary
+ * @tparam T
+ * @param strings
+ * @return
+ */
+template <typename T>
+std::vector<math::SizeType> W2VLoader<T>::StringsToIndices(std::vector<std::string> const &strings)
+{
+  std::vector<SizeType> indexes;
+  if (strings.size() >= 2 * window_size_ + 1)  // Don't bother processing too short inputs
+  {
+    indexes = vocab_.PutSentenceInVocab(strings);
+  }
+  return indexes;
+}
+
+/**
+ * Preprocesses a string turning it into a vector of words
+ * @tparam T
+ * @param s
+ * @return
+ */
+template <typename T>
+std::vector<std::string> W2VLoader<T>::PreprocessString(std::string const &s)
+{
+  std::string result;
+  result.reserve(s.size());
+  for (auto const &c : s)
+  {
+    result.push_back(std::isalpha(c) != 0 ? static_cast<char>(std::tolower(c)) : ' ');
+  }
+
+  std::string              word;
+  std::vector<std::string> words;
+  for (std::stringstream ss(result); ss >> word;)
+  {
+    words.push_back(word);
+  }
+  return words;
+}
+
+template <typename T>
+void W2VLoader<T>::UpdateCursor()
+{
+  if (this->mode_ != DataLoaderMode::TRAIN)
+  {
+    throw exceptions::InvalidMode("mode not supported.");
+  }
+}
+
+template <typename T>
+bool W2VLoader<T>::IsModeAvailable(DataLoaderMode mode)
+{
+  return mode == DataLoaderMode::TRAIN;
+}
+
+template <typename T>
+T const W2VLoader<T>::WindowContextUnused = static_cast<T>(-1);
+
+///////////////////////////////
+/// EXPLICIT INSTANTIATIONS ///
+///////////////////////////////
+
+template class W2VLoader<float>;
```

Comment:
no integral types?

Reply:
Yes, because there is static assert in class declaration
public:
  static_assert(fetch::meta::IsFloat<T> || math::meta::IsFixedPoint<T>,
                "The intended T is the typename for the data input to the neural network, which "
                "should be a float or double or fixed-point type.");

Reply:
i don't think that's true anymore though, right?
I mean, there may be many pieces that won't work well or need adjustment for full integral compatibility, but we can at least get everything integral compilable right?